Samples/CayenneDecoder/Classes/CayenneDecoder.cs

```diff
@@ -10,218 +10,149 @@ public class CayenneDecoder
-            return false;
-        }
-
-        public bool TryGetGyrometer(out Gyrometer gyrometer)
-        {
-            var ret = GetOutValues(CayenneTypes.Gyrometer);
-            if (ret != null)
-            {
-                gyrometer = new Gyrometer()
-                {
-                    Channel = ret[0],
-                    X = ((Int16)(ret[1] << 8) + ret[2]) / 100.0,
-                    Y = ((Int16)(ret[3] << 8) + ret[4]) / 100.0,
-                    Z = ((Int16)(ret[5] << 8) + ret[6]) / 100.0
-                };
-                return true;
-            }
-            gyrometer = null;
-            return false;
-        }
-
-        public bool TryGetGPSLocation(out GPSLocation gPSLocation)
-        {
-            var ret = GetOutValues(CayenneTypes.Gps);
-            if (ret != null)
-            {
-                gPSLocation = new GPSLocation();
-
-                gPSLocation.Channel = ret[0];
-                double sign = 1.0;
-                if ((ret[1] & 0x80) == 0x80)
-                {
-                    ret[1] = (byte)(ret[1] & 0x7F);
-                    sign = -1.0;
-                }
-                gPSLocation.Latitude = sign * ((ret[1] << 16) + (ret[2] << 8) + ret[3]) / 10000.0;
-                sign = 1.0;
-                if ((ret[4] & 0x80) == 0x80)
-                {
-                    ret[4] = (byte)(ret[4] & 0x7F);
-                    sign = -1.0;
-                }
-                gPSLocation.Longitude = sign * ((ret[4] << 16) + (ret[5] << 8) + ret[6]) / 10000.0;
-                sign = 1.0;
-                if ((ret[7] & 0x80) == 0x80)
-                {
-                    ret[7] = (byte)(ret[7] & 0x7F);
-                    sign = -1.0;
-                }
-                gPSLocation.Altitude = sign * ((ret[7] << 16) + (ret[8] << 8) + ret[9]) / 100.0;
-                return true;
-            }
-            gPSLocation = null;
-            return false;
-        }
-
-        private byte[] GetOutValues(CayenneTypes cayenneTypes)
-        {
-            byte[] outArray = null;
             byte channel = 0;
             int cursor = 0;
             try
             {
-                while (cursor != buffer.Length)
+                foreach (var cayenneTypes in Enum.GetValues(typeof(CayenneTypes)))
                 {
-                    channel = buffer[cursor++];
-                    var type = buffer[cursor++];
-                    var size = (byte)Enum.Parse<CayenneTypeSize>(Enum.GetName(typeof(CayenneTypes), type));
-                    if (type == (byte)cayenneTypes)
+                    while (cursor != buffer.Length)
                     {
-                        // found what we want to decode
-                        outArray = new byte[size - 1];
-                        outArray[0] = channel;
-                        Array.Copy(buffer, cursor, outArray, 1, outArray.Length - 1);
-                        return outArray;
+                        channel = buffer[cursor++];
```

Comment:
Suggest: you could move the logic of creating the concrete Cayenne type to each constructor, so this method is cleaner:
```c#
new DigitalInput(channel, value.Span(cursor, size));
or
new Accelerator(channel, value.Span(cursor, size));
```

And have constructors like:
```c#
public Accelerator(byte channel, ReadOnlySpan<byte> buffer)
{
    this.X = ((Int16)(buffer[0] << 8) + buffer[1]) / 1000.0;
    this.Y = ((Int16)(buffer[2] << 8) + buffer[3]) / 1000.0;
    this.Z = ((Int16)(buffer[4] << 8) + buffer[5]) / 1000.0
}
```

Reply:
You could even get rid of the CayenneTypeSize, by having each type have a property with the Byte length.
In each constructor you check if the span has the minimum expected size, if not throw an exception.

Reply:
I think that in terms of architecture, if we want or need at some point to use the encoder class as well to generate the byte arays, I would leave the conversion part of the converter class and have the device objects as light as possible. I was hesitating to do it like you proposed and finally decided to leave all in the conversion class.
Having an enum class for the size makes it easier to avoid mistake with "magic numbers". because the way to get rid of them is to hard code them in the switch section.